I2C writeto_then_readfrom in a single transaction on u2if

ryanmesh-sd opened the pull request:
Implemented WRITE_THEN_READ on I2C interface in U2IF (associated PR: https://github.com/adafruit/u2if/pull/12)

This allows reading from specific registers in I2C devices without sending two separate commands from the host, which significantly increases latency between the pointer write and read and causes timeouts in certain I2C devices (like the ones I've been using).

ryanmesh-sd committed:
Added WRITE_THEN_READ command for u2if

ryanmesh-sd committed:
Formatted rp2040_u2if.py with Black formatter

Previous commit failed CI due to changes made by formatter.

ryanmesh-sd commented:
Looks to now pass tests on my branch CI. Would it be better to do this with a fallback to the old (two separate HID reports) method to accommodate U2IF devices using older firmware?

FoamyGuy commented:
> Looks to now pass tests on my branch CI. Would it be better to do this with a fallback to the old (two separate HID reports) method to accommodate U2IF devices using older firmware?

Yes, I think it would be best for this to have fallback for devices with the older firmware. 

Thanks for working on this!

ryanmesh-sd committed:
Allow fallback on RP2040 U2IF WRITE_THEN_READ to old write and read functions in series
Allows usage on older versions of U2IF firmware

ryanmesh-sd committed:
Merge branch 'adafruit:main' into main

ryanmesh-sd commented:
Hi @FoamyGuy, I've added the fallback. Let me know what you think!

ryanmesh-sd committed:
Move RP2040 U2IF I2C fallback flag to init

ryanmesh-sd committed:
Merge branch 'main' of https://github.com/ryanmesh-sd/Adafruit_Blinka

ryanmesh-sd committed:
Merge branch 'adafruit:main' into main

makermelissa-piclaw committed:
Clean up I2C write_then_read implementation

- Fix end_w/end_r falsy check: use 'is not None' to handle end=0 correctly
- Differentiate error message for write_then_read vs regular write
- Remove commented-out debug print statements
